README.md

```diff
@@ -291,13 +291,28 @@ const instance = new Test();
 type TestTypeFromInstance = Class<typeof instance>;
 ```
 
+This works with type parameters:
+```js
+function newInstance<T>(c: Class<T>) {
+    ...
+}
+```
+
 ### TypeScript
 
 ```ts
 class Test {};
 type TestType = typeof Test;
 ```
 
+This unfortunately doesn't work with type parameters:
+```ts
+// error: 'T' only refers to a type, but is being used as a value here.(2693)
+function newInstance<T>(c: typeof T) {
```

Comment:
But you can:
```ts

declare function newInstance<T>(c: new (..._: any) => T): T;

var x = newInstance(class X { y = 1; });
x.y = 2
```

Reply:
Okay, I could include that, but I bet it has some limitations like not having type information about the static members of the class, right?

Reply:
Well, It's not clear how you want to use it. Flow has some limitations too, it's not like that in all scenarios you just get the `static`s part with this annotation. It's still a generic, (but the last Flow version I've used is 0.85). But, TS is anyway not the best in inference of `static` properties and even when it's suppose to have the information.

```ts
declare class Hi {
    static sss(): number;
    iii(): string;
}

var h = new Hi();

h.constructor.sss();
//            ^^^ <- Error in TS - Works in Flow
```

Reply:
This case works in Flow at least:
```js
declare class Hi {
    static sss(): number;
    iii(): string;
}

function foo<T: Hi>(cls: Class<T>) {
	cls.sss() // no error
        cls.aaa() // Cannot call `cls.aaa` because property `aaa` is missing in statics of `Hi`
}
```

I'm just saying if we include a note about `new (..._: any) => T` types, we should mention that it definitely doesn't capture static type information, but that `Class<T>` can.

Reply:
In the moment you add `:` in the generics it's become pointless. It's not a "real life usage". Because I would probably write it as:
```ts
declare class Hi {
    static sss(): number;
    iii(): string;
}


function foo(cls: typeof Hi) {
    cls.sss() // no error
    cls.aaa() // Cannot call `cls.aaa` because property `aaa` is missing in statics of `Hi`
}
```

I think that this demo is quite pointless. You better right about the fact that TS is really weak about static fields and you cannot get them through `constructor`, aka:
```ts
const h = new Hi();
h.constructor.sss(); // <- Error, doesn't know `sss` exists
```

But `Class<T>` utility isn't that different than `typeof`. The fact TS doesn't recognize `static` fields is the real different.
If you wouldn't use `:` in the demo, only if a function isn't exported it might work and than you wouldn't need any type annotations (this is where Flow shines).

Reply:
By the way, with time, I can definitely say that me (and I guess a lots of other) tends to "over-type" with complex types when usually the simple annotation probably be more correct and work better

Reply:
> In the moment you add `:` in the generics it's become pointless
I think you just haven't encountered a situtation where this matters, because here is a real world use case from my own code, using the flow type defs for `sequelize` (which I adapted from the TypeScript definitions people had created).
```js

export default async function simpleBulkUpsert<
  TAttributes: Object,
  TInitAttributes: Object,
  TInstance: Model<TAttributes, TInitAttributes>
>({
  model,
  values,
  indexCols,
  options = {},
}: {
  model: Class<TInstance>,
  values: $ReadOnlyArray<TInitAttributes>,
  indexCols: $ReadOnlyArray<$Keys<TAttributes>>,
  options: UpsertOptions<TAttributes>,
}): Promise<Array<TInstance>> {
  const { sequelize, QueryGenerator } = model
  const uniqueValues = new Map()
  const specifiedAttrs = {}
  values.forEach((value: TInitAttributes) => {
    const key = JSON.stringify(pick(value, indexCols))
    uniqueValues.set(key, value)
    Object.keys(value).forEach((name: $Keys<TInitAttributes>) => {
      if (value.hasOwnProperty(name))
        specifiedAttrs[name] = model.attributes[name]
    })
  })
  ...
```
In this case, the `model` passed in is a subclass of `Sequelize.Model`, and it has a `static attributes` field being accessed in this function.

Reply:
I'm not trying to poo-poo TypeScript or anything, I'm in the process of switching to it due to numerous struggles with Flow in other areas.  I just think this is an honest shortcoming of TypeScript that other users will benefit from understanding here.

Reply:
Certainly you could argue that the sequelize types above are "over-typed", but the alternative is less type-safe, so it's only a subjective argument that the static class member type checking in my example is useless.  The goal of this repo is to present objective facts about the differences between Flow and TypeScript, not subjective pros and cons.  I guess I could remove some of the subjective language in my PR ("unfortunately") though.

Reply:
- I didn't talk about TS vs Flow. I've gave examples where Flow is better than TS
- Overtype, wasn't said about comparing to "losing needed type information", was about doing sophisticated types that aren't give any type safety, only use-less meta data
- I don't say `sequelize` definitions are over-typed (By the way I guess in TS it would done very differently). I said giving overtype examples for non-real life senecrious isn't good a thing.

The main point, I think `Class<T>` vs `typeof T` isn't a good example, because it's misleading and unclear to the point, But showing TS can't handle `static` properties or  "prototype chain" is way more relevant while Flow does it

Reply:
Okay, I'm just having trouble understanding why you're against my proposed changes.  Would you be in agreement with something like

> The TypeScript equivalent of `Class<T>` is `typeof T` in cases where `T` is a value, and `new (..._: any) => T` in cases where `T` is a type; however, unlike `Class<T>` this latter case doesn't include static class member types.

Reply:
Okay for reference with latest typescript playground:

```ts
class Foo {
  static bar: number = 2
}

const Typeof: typeof Foo = Foo
const a: string = Typeof.bar

const New: (new (..._: any) => Foo) = Foo
const b: string = New.bar
```

```
const a: string
Type 'number' is not assignable to type 'string'.(2322)

Property 'bar' does not exist on type 'new (..._: any) => Foo'.(2339)
```
So I think anything we document about differences in class typing must discuss both `typeof T` and `new (..._: any) => T`, and the differences between each.

Reply:
TS don't deal `static`s as part of prototype object well. `typeof SomeClass` will work as much as `typeof anyInstanceObject`. But `anyInstanceObject.prototype.constructor.*` won't work in TS, Only in Flow.

I didn't say I'm against, I'm not a maintainer of this project and don't decide what goes in or not.

I think it's misleading to present it as "`Class<T>` vs `typeof T`". It's not the important part for my opinion and quite misleading. Talking about limitation to extract `static` properties from TS comparing to Flow is way more important and accurate

Reply:
It's very important for people migrating from Flow to TS to learn what to use instead of `Class<T>`.  They need to know that they won't be able to fully replicate `Class<T>` in some situations.

Our experience is obviously very different, as I've never had a need to use `anyInstanceObject.prototype.constructor.*` in code.  Trust me, `Class<T>` vs `typeof T`/`new (..._: any) => T` is not misleading or unimportant for people who have used `Class<T>` heavily migrating to TS.